whois-nrtm4-client/src/main/java/net/ripe/db/nrtm4/client/importer/DeltaImporter.java

```diff
@@ -0,0 +1,175 @@
+                return;
+            }
+
+            final String[] deltaFileResponse = StringUtils.split(new String(deltaFilePayload, StandardCharsets.UTF_8), RECORD_SEPARATOR);
+
+            final Metadata metadata = getMetadata(deltaFileResponse);
+
+            final List<MirrorDeltaInfo> mirrorObjectInfos = getMirrorDeltaObjects(deltaFileResponse);
+
+            if (!delta.getHash().equals(calculateSha256(deltaFilePayload))){
+                LOGGER.error("Snapshot hash doesn't match, skipping import");
+                return;
+            }
+
+            if (!metadata.sessionId().equals(updateNotificationFile.getSessionID())){
+                LOGGER.error("The session is not the same in the UNF and snapshot");
+                return;
+            }
+
+            mirrorObjectInfos.forEach(this::applyDeltaRecord);
+
+            nrtm4ClientInfoRepository.saveDeltaFileVersion(source, metadata.version, metadata.sessionId());
+        });
+    }
+
+    private void applyDeltaRecord(final MirrorDeltaInfo deltaInfo){
+        if (deltaInfo.getAction().equals(MirrorDeltaInfo.Action.ADD_MODIFY)){
+
+            final RpslObjectUpdateInfo rpslObjectUpdateInfo = nrtm4ClientRepository
+                    .getMirroredObjectId(deltaInfo.getRpslObject().getType(), deltaInfo.getRpslObject().getKey().toString());
+            if (rpslObjectUpdateInfo == null) {
+                applyDeltaCreation(deltaInfo);
+                return;
+            }
+            applyDeltaUpdate(deltaInfo, rpslObjectUpdateInfo);
+        } else {
+            applyDeltaDeletion(deltaInfo);
+        }
+    }
+
+    private void applyDeltaCreation(final MirrorDeltaInfo deltaInfo) {
+        final RpslObjectUpdateInfo rpslObjectCretedInfo = nrtm4ClientRepository.persistRpslObject(deltaInfo.getRpslObject());
+        nrtm4ClientRepository.createIndexes(deltaInfo.getRpslObject(), rpslObjectCretedInfo);
+    }
+
+    private void applyDeltaDeletion(final MirrorDeltaInfo deltaInfo) {
+        final RpslObjectUpdateInfo rpslObjectUpdateInfo = nrtm4ClientRepository.getMirroredObjectId(deltaInfo.getObjectType(), deltaInfo.getPrimaryKey());
+
+        if (serialDoesNotExist(rpslObjectUpdateInfo)){
+            LOGGER.error("delta with pkey: {} not deleted because serial doesn't exist", deltaInfo.getPrimaryKey());
+            return;
+        }
+        nrtm4ClientRepository.removeMirroredObjectAndUpdateSerials(rpslObjectUpdateInfo);
+    }
+
+    private void applyDeltaUpdate(final MirrorDeltaInfo deltaInfo, final RpslObjectUpdateInfo rpslObjectUpdateInfo) {
+        if (serialDoesNotExist(rpslObjectUpdateInfo)) {
+            LOGGER.error("delta with pkey: {} not updated because serial doesn't exist", deltaInfo.getPrimaryKey());
+            return;
+        }
+        nrtm4ClientRepository.updateMirroredObject(deltaInfo.getRpslObject(), rpslObjectUpdateInfo);
+    }
+
+    private boolean serialDoesNotExist(final RpslObjectUpdateInfo rpslObjectUpdateInfo) {
```

Comment:
This logic is hard to understand, e.g. when `serialDoesNotExist == false` 🤔  Can you flip the logic to `serialExists` it's easier to read